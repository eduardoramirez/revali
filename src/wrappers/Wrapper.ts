import { AnyConstructor } from '../types';
import {
  GraphQLNonNull,
  GraphQLScalarLiteralParser,
  GraphQLScalarType,
  GraphQLScalarValueParser,
  GraphQLType,
  isNonNullType,
} from 'graphql';

export type Wrapper<T, G extends GraphQLType = GraphQLType> = {
  graphQLType: G,
  type: T,
  nullable?: boolean,
};

export type WrapperOrType<T, G extends GraphQLType = GraphQLType> = Wrapper<T, G> | AnyConstructor<T>;

export interface TypedGraphQLScalar<TInternal> extends GraphQLScalarType {
  parseValue: GraphQLScalarValueParser<TInternal>;
  parseLiteral: GraphQLScalarLiteralParser<TInternal>;
}

export const isWrapper = <T, G extends GraphQLType>(x: WrapperOrType<T, G>): x is Wrapper<T, G> => {
  return 'graphQLType' in x && 'type' in x;
};

export function resolveWrapper<T, G extends GraphQLType>(wrapper: Wrapper<T, G>, dontWrap: true): G;
export function resolveWrapper<T, G extends GraphQLType>(wrapper: Wrapper<T, G>, dontWrap: false): G | GraphQLNonNull<G>;
export function resolveWrapper<T, G extends GraphQLType>(wrapper: Wrapper<T, G>): G | GraphQLNonNull<G>;
export function resolveWrapper<T, G extends GraphQLType>(wrapper: Wrapper<T, G>, dontWrap?: boolean): G | GraphQLNonNull<G> {
  const type = wrapper.graphQLType;
  return (wrapper.nullable || dontWrap)
    ? type
    : new GraphQLNonNull(type);
}

export const wrapScalar = <T>(scalar: TypedGraphQLScalar<T>): Wrapper<T, GraphQLScalarType> => {
  return {
    graphQLType: scalar,
    type: (null as any) as T,
  };
};

export const unsafeWrapType = <T extends GraphQLType>(type: T): Wrapper<any, T> => {
  return {
    graphQLType: type,
    type: null as any,
  };
};
