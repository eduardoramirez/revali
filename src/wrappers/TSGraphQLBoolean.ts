import { GraphQLBoolean } from 'graphql';
import { wrapScalar } from './Wrapper';

export default wrapScalar<boolean>(GraphQLBoolean);
