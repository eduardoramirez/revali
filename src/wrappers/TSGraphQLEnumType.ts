import { Wrapper } from './Wrapper';
import { GraphQLEnumType, GraphQLNonNull, GraphQLType } from 'graphql';
import { mapValues } from 'lodash';

export type TSGraphQLEnumTypeConfig<K> = {
  name: string;
}

export default class TSGraphQLEnumType<K extends string, TEnum> implements Wrapper<K, GraphQLEnumType> {
  graphQLType: GraphQLEnumType;
  type: K;
  constructor(enumType: Record<K, TEnum>, config: TSGraphQLEnumTypeConfig<K>) {
    this.graphQLType = new GraphQLEnumType({
      ...config,
      values: mapValues(enumType, (value) => ({
        value,
      })),
    });
    this.type = null as any as K;
  }
}
