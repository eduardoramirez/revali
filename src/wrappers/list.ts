import { isWrapper, Wrapper, WrapperOrType } from './Wrapper';
import { GraphQLList } from 'graphql';
import { getType } from '../typeHelpers';

export default function list<T>(type: WrapperOrType<T>): Wrapper<T[], GraphQLList<any>> {
  const currentType = getType(type, true);
  const transformOutput = isWrapper(type) && type.transformOutput;
  return {
    graphQLType: new GraphQLList(currentType),
    transformOutput: transformOutput
      ? (values: T[]) => values.map(transformOutput.bind(type))
      : undefined,
    type: [],
  }
}
