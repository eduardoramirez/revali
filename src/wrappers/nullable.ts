import { graphQLTypeForWrapper, Wrapper, WrapperOrType } from './Wrapper';
import { GraphQLNonNull } from 'graphql';

function nullable<T>(type: WrapperOrType<T>): Wrapper<T | null> {
  const currentType = graphQLTypeForWrapper(type);
  let graphQLType = currentType;
  if (currentType instanceof GraphQLNonNull) {
    graphQLType = currentType.ofType;
  }
  return {
    graphQLType,
    type: null,
  };
}
