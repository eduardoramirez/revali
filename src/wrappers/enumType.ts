import constantCase from 'constant-case'
import {GraphQLEnumType, GraphQLEnumValueConfigMap} from 'graphql'
import pascalCase from 'pascal-case'

import {Wrapper} from 'revali/wrappers/Wrapper'

export enum EnumTypeCase {
  Pascal,
  Constant,
}

export interface EnumValueConfig {
  deprecationReason?: string
  description?: string
}

export type EnumValueConfigMap<K extends keyof any> = {[key in K]?: EnumValueConfig}

export interface EnumTypeConfig<K extends keyof any> {
  name: string
  description?: string
  changeCase?: EnumTypeCase
  additional?: EnumValueConfigMap<K>
}

const performChangeCase = (type: EnumTypeCase, value: string): string => {
  switch (type) {
    case EnumTypeCase.Constant:
      return constantCase(value)
    case EnumTypeCase.Pascal:
      return pascalCase(value)
  }
}

export function enumType<K extends string, TEnum extends string | number>(
  enumObject: Record<K, TEnum>,
  config: EnumTypeConfig<K>
): Wrapper<TEnum, GraphQLEnumType> {
  const {changeCase} = config

  const getKey = (key: string) => (changeCase ? performChangeCase(changeCase, key) : key)

  const graphQLType = new GraphQLEnumType({
    ...config,
    values: Object.keys(enumObject).reduce((map: GraphQLEnumValueConfigMap, key) => {
      if (!isNaN(parseInt(key, 10))) {
        return map
      }
      return {
        ...map,
        [getKey(key)]: {
          ...(config.additional && (config.additional as any)[key]),
          value: (enumObject as any)[key],
        },
      }
    }, {}),
  })

  return {
    graphQLType,
    type: (null as any) as TEnum,
  }
}
