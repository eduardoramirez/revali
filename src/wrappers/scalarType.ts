import {GraphQLScalarType, GraphQLScalarTypeConfig} from 'graphql'

import {Wrapper} from 'revali/wrappers/Wrapper'

export function scalarType<TInternal, TExternal>(
  config: GraphQLScalarTypeConfig<TInternal, TExternal>
): Wrapper<TInternal, GraphQLScalarType> {
  return {
    graphQLType: new GraphQLScalarType(config),
    type: (null as any) as TInternal,
  }
}
