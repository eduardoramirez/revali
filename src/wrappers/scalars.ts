import {GraphQLBoolean, GraphQLFloat, GraphQLID, GraphQLInt, GraphQLString} from 'graphql'

import {wrapScalar} from 'revali/wrappers/Wrapper'

export const TSGraphQLBoolean = wrapScalar<boolean>(GraphQLBoolean)
export const TSGraphQLFloat = wrapScalar<number>(GraphQLFloat)
export const TSGraphQLID = wrapScalar<number | string>(GraphQLID)
export const TSGraphQLInt = wrapScalar<number>(GraphQLInt)
export const TSGraphQLString = wrapScalar<string>(GraphQLString)
