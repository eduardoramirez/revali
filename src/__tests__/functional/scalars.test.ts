import {
  getIntrospectionQuery,
  graphql,
  GraphQLScalarType,
  GraphQLSchema,
  printSchema,
} from 'graphql'

import {
  Arg,
  compileSchema,
  Field,
  ObjectType,
  scalarType,
  TSGraphQLBoolean,
  TSGraphQLFloat,
  TSGraphQLID,
  TSGraphQLInt,
  TSGraphQLString,
} from 'revali/index'
import {wrapScalar} from 'revali/wrappers/Wrapper'

const CustomScalar = scalarType<string, string>({
  name: 'Custom',
  description: 'My custom scalar wohoo',
  parseLiteral: () => 'did parseLiteral',
  parseValue: () => 'did parseValue',
  serialize: () => 'did serialize',
})

const ObjectScalar = new GraphQLScalarType({
  name: 'ObjectScalar',
  parseLiteral: () => ({
    value: 'did parseLiteral',
  }),
  parseValue: () => ({
    value: 'did parseValue',
  }),
  serialize: obj => obj.value,
})

class CustomObjectType {
  public value!: string
}
const ObjectScalarType = wrapScalar<CustomObjectType>(ObjectScalar)

describe('Scalars', () => {
  let schema: GraphQLSchema
  let scalarArg: any

  beforeAll(() => {
    @ObjectType()
    class SampleObject {
      @Field({type: TSGraphQLID})
      public idField!: string

      @Field()
      public implicitFloatField!: number

      @Field({type: TSGraphQLFloat})
      public explicitFloatField!: number

      @Field({type: TSGraphQLInt})
      public intField!: number

      @Field()
      public implicitStringField!: string

      @Field({type: TSGraphQLString})
      public explicitStringField!: string

      @Field()
      public implicitBooleanField!: boolean

      @Field({type: TSGraphQLBoolean})
      public explicitBooleanField!: boolean

      @Field({type: CustomScalar})
      public customScalarField!: string
    }

    @ObjectType()
    class Query {
      @Field({type: SampleObject})
      public mainQuery(): SampleObject {
        return {} as any
      }

      @Field({type: CustomScalar})
      public returnScalar(): string {
        return 'returnScalar'
      }

      @Field({type: TSGraphQLBoolean, arg: CustomScalar})
      public argScalar(@Arg('scalar') scalar: string): boolean {
        scalarArg = scalar
        return true
      }

      @Field({type: TSGraphQLBoolean, arg: ObjectScalarType})
      public objectArgScalar(@Arg('scalar') scalar: CustomObjectType): boolean {
        scalarArg = scalar
        return true
      }
    }

    schema = compileSchema({Query})
  })

  afterEach(() => {
    scalarArg = undefined
  })

  it('generates scalars correctly', async () => {
    const {errors} = await graphql(schema, getIntrospectionQuery())
    expect(errors).toBeFalsy()
    expect(printSchema(schema)).toMatchSnapshot()
  })

  it('should properly serialize data', async () => {
    const query = `query {
        returnScalar
      }`
    const result = await graphql(schema, query)
    const returnScalar = result.data!.returnScalar

    expect(returnScalar).toEqual('did serialize')
  })

  it('should properly parse args', async () => {
    const query = `query {
        argScalar(scalar: "test")
      }`

    const result = await graphql(schema, query)

    expect(result).toHaveProperty('data', {
      argScalar: true,
    })
    expect(result.errors).toBeFalsy()
    expect(scalarArg!).toEqual('did parseLiteral')
  })

  it('should properly parse scalar object', async () => {
    const query = `query {
        objectArgScalar(scalar: "test")
      }`
    await graphql(schema, query)

    expect(scalarArg!).toEqual({value: 'did parseLiteral'})
  })
})
