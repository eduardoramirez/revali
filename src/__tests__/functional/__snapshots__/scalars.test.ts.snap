// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scalars generates scalars correctly 1`] = `
"\\"\\"\\"My custom scalar wohoo\\"\\"\\"
scalar Custom

scalar ObjectScalar

type Query {
  mainQuery: SampleObject!
  returnScalar: Custom!
  argScalar(scalar: Custom!): Boolean!
  objectArgScalar(scalar: ObjectScalar!): Boolean!
}

type SampleObject {
  idField: ID!
  implicitFloatField: Float!
  explicitFloatField: Float!
  intField: Int!
  implicitStringField: String!
  explicitStringField: String!
  implicitBooleanField: Boolean!
  explicitBooleanField: Boolean!
  customScalarField: Custom!
}
"
`;
